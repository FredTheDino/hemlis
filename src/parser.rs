#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    macro_rules! gen_parser {
        ($a:ident, $p:ident) => {
            pub fn $a<'t>(src: &'t str) -> String {
                pub fn inner<'t>(src: &'t str) -> Result<String, String> {
                    use crate::{ast::Ast, grammer, lexer};
                    use std::io::BufWriter;

                    let parser = grammer::$p::new();
                    let out = parser
                        .parse(
                            lexer::lex(&src)
                                .into_iter()
                                .map(|(token, span)| Ok((span.start, token?, span.end))),
                        )
                        .map_err(|x| format!("{:?}", x))?;

                    let mut buf = BufWriter::new(Vec::new());
                    out.show(0, &mut buf).map_err(|x| format!("{:?}", x))?;
                    let inner = buf.into_inner().map_err(|x| format!("{:?}", x))?;
                    String::from_utf8(inner).map_err(|x| format!("{:?}", x))
                }
                match inner(src) {
                    Ok(s) | Err(s) => s,
                }
            }
        };
    }

    gen_parser!(p_header, HeaderParser);
    gen_parser!(p_import, ImportParser);

    #[test]
    fn empty_string() {
        assert_snapshot!(p_header(""));
    }

    #[test]
    fn normal_definition() {
        assert_snapshot!(p_header("module A (a, b, c) where"));
    }

    #[test]
    fn everything_header() {
        assert_snapshot!(p_header(r#"
module A.B.C (a, class B, C, D(..), E(F, G), 
 (+), type (+), module H) where

import A as A
import A.B.C as A.C
import A.B.C (a, class B, 
 C, D(..), E(F, G), (+), type (+))
import A.B.C hiding (foo)
        "#));
    }

    #[test]
    fn simple_import() {
        assert_snapshot!(p_import("import A as A"))
    }

    #[test]
    fn import_with_newlines() {
        assert_snapshot!(p_import("import A (foo\n , bar\n , baz)"))
    }
}
