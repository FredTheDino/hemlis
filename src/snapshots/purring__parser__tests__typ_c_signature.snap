---
source: src/parser.rs
expression: "p_typ(\"\n     forall out pkA pkB propsB rowB propsA rowA select\n   . GetPk (C pkA N) propsA\n  => ReflectArray pkA\n  => HasMagicMapping rowA rowA\n  => AllColumns rowA select\n  => HasMagicSelect propsA rowA select select out\n  => GetPk (C pkB N) propsB\n  => ReflectArray pkB\n  => HasMagicMapping rowB rowB\n  => Table propsA rowA\n  -> Table propsB rowB\n  -> (Record out -> Record rowB)\n  -> Sql Unit\")"
---
76 of 76
===
Module
 Header
  QProperName
   NULL
   ProperName
    7..10
  Decl::Sig
   Name
    18..19
   Typ::Forall
     TypVarBinding
      Name
       36..39
      NULL
     TypVarBinding
      Name
       40..43
      NULL
     TypVarBinding
      Name
       44..47
      NULL
     TypVarBinding
      Name
       48..54
      NULL
     TypVarBinding
      Name
       55..59
      NULL
     TypVarBinding
      Name
       60..66
      NULL
     TypVarBinding
      Name
       67..71
      NULL
     TypVarBinding
      Name
       72..78
      NULL
    Typ::Constrained
     Constraint
      QProperName
       NULL
       ProperName
        84..89
       Typ::App
        Typ::App
         Typ::Constructor
          QProperName
           NULL
           ProperName
            91..92
         Typ::Var
          Name
           93..96
        Typ::Constructor
         QProperName
          NULL
          ProperName
           97..98
       Typ::Var
        Name
         100..106
     Typ::Constrained
      Constraint
       QProperName
        NULL
        ProperName
         112..124
        Typ::Var
         Name
          125..128
      Typ::Constrained
       Constraint
        QProperName
         NULL
         ProperName
          134..149
         Typ::Var
          Name
           150..154
         Typ::Var
          Name
           155..159
       Typ::Constrained
        Constraint
         QProperName
          NULL
          ProperName
           165..175
          Typ::Var
           Name
            176..180
          Typ::Var
           Name
            181..187
        Typ::Constrained
         Constraint
          QProperName
           NULL
           ProperName
            193..207
           Typ::Var
            Name
             208..214
           Typ::Var
            Name
             215..219
           Typ::Var
            Name
             220..226
           Typ::Var
            Name
             227..233
           Typ::Var
            Name
             234..237
         Typ::Constrained
          Constraint
           QProperName
            NULL
            ProperName
             243..248
            Typ::App
             Typ::App
              Typ::Constructor
               QProperName
                NULL
                ProperName
                 250..251
              Typ::Var
               Name
                252..255
             Typ::Constructor
              QProperName
               NULL
               ProperName
                256..257
            Typ::Var
             Name
              259..265
          Typ::Constrained
           Constraint
            QProperName
             NULL
             ProperName
              271..283
             Typ::Var
              Name
               284..287
           Typ::Constrained
            Constraint
             QProperName
              NULL
              ProperName
               293..308
              Typ::Var
               Name
                309..313
              Typ::Var
               Name
                314..318
            Typ::Arr
             Typ::App
              Typ::App
               Typ::Constructor
                QProperName
                 NULL
                 ProperName
                  324..329
               Typ::Var
                Name
                 330..336
              Typ::Var
               Name
                337..341
             Typ::Arr
              Typ::App
               Typ::App
                Typ::Constructor
                 QProperName
                  NULL
                  ProperName
                   347..352
                Typ::Var
                 Name
                  353..359
               Typ::Var
                Name
                 360..364
              Typ::Arr
               Typ::Arr
                Typ::App
                 Typ::Constructor
                  QProperName
                   NULL
                   ProperName
                    371..377
                 Typ::Var
                  Name
                   378..381
                Typ::App
                 Typ::Constructor
                  QProperName
                   NULL
                   ProperName
                    385..391
                 Typ::Var
                  Name
                   392..396
               Typ::App
                Typ::Constructor
                 QProperName
                  NULL
                  ProperName
                   403..406
                Typ::Constructor
                 QProperName
                  NULL
                  ProperName
                   407..411
===
